plugins/query-insights/src/main/java/org/opensearch/plugin/insights/core/service/QueryInsightsService.java

```diff
@@ -0,0 +1,282 @@
+    /**
+     * window start timestamp for each top queries collector
+     */
+    private final Map<MetricType, Long> topQueriesWindowStart = new HashMap<>();
+
+    /**
+     * Create the TopQueriesByLatencyService Object
+     *
+     * @param threadPool     The OpenSearch thread pool to run async tasks
+     */
+    @Inject
+    public QueryInsightsService(ThreadPool threadPool) {
+        this.topQueriesByLatencyStore = new PriorityBlockingQueue<>(
+            topNSize,
+            Comparator.comparingLong(a -> a.getMeasurement(MetricType.LATENCY).getValue().longValue())
+        );
+        this.topQueriesByCpuStore = new PriorityBlockingQueue<>(
+            topNSize,
+            Comparator.comparingDouble(a -> a.getMeasurement(MetricType.CPU).getValue().doubleValue())
+        );
+        this.topQueriesByJvmStore = new PriorityBlockingQueue<>(
+            topNSize,
+            Comparator.comparingDouble(a -> a.getMeasurement(MetricType.JVM).getValue().doubleValue())
+        );
+        this.threadPool = threadPool;
+    }
+
+    private PriorityBlockingQueue<SearchQueryRecord> getStoreByMetricType(MetricType type) {
+        if (type == MetricType.LATENCY) {
+            return topQueriesByLatencyStore;
+        } else if (type == MetricType.CPU) {
+            return topQueriesByCpuStore;
+        } else if (type == MetricType.JVM) {
+            return topQueriesByJvmStore;
+        } else {
+            throw new IllegalArgumentException(
+                "Invalid metricType [" + type + "]" + " should be one of [" + MetricType.allMetricTypes() + "]"
+            );
+        }
+    }
+
+    /**
+     * Ingest the query data into local stores
+     *
+     * @param metricType {@link MetricType}
+     * @param record the record to ingest
+     */
+    public void addRecord(final MetricType metricType, SearchQueryRecord record) {
+        PriorityBlockingQueue<SearchQueryRecord> store = getStoreByMetricType(metricType);
+        this.threadPool.schedule(() -> {
+            checkAndResetWindow(metricType, record.getTimestamp());
+            if (record.getTimestamp() > topQueriesWindowStart.get(metricType)) {
+                store.add(record);
+                // remove top elements for fix sizing priority queue
+                if (store.size() > this.getTopNSize()) {
+                    store.poll();
+                }
+            }
+        }, delay, ThreadPool.Names.GENERIC);
+
+        log.debug(String.format(Locale.ROOT, "successfully ingested record type: %s", metricType));
+    }
+
+    private synchronized void checkAndResetWindow(MetricType metricType, Long timestamp) {
+        Long windowStart = calculateWindowStart(timestamp);
+        // reset window if the current window is outdated
+        if (!topQueriesWindowStart.containsKey(metricType) || topQueriesWindowStart.get(metricType) < windowStart) {
+            getStoreByMetricType(metricType).clear();
```

Comment:
isn't it a rolling window for last x hours?

Reply:
Rolling window for last x hour may require us to record all search request records. But based on our discussion, I will refacor the code to avoid lossing records at the end of one window.